ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/Round.java

```diff
@@ -0,0 +1,139 @@
+This means:
+
+ROUND(1.1) -> 1
+ROUND(1.5) -> 2
+ROUND(1.9) -> 2
+
+ROUND(-1.1) -> -1
+ROUND(-1.5) -> -1    Note this is not -2! We round up and up is in a positive direction.
+ROUND(-1.9) -> -2
+ */
+@UdfDescription(name = "Round", description = Round.DESCRIPTION)
+public class Round {
+
+  static final String DESCRIPTION =
+      "Returns the value rounded to the specified number of decimal places. "
+      + "If the number of decimal places is omitted it defaults to zero."
+      + "Numbers are rounded to the nearest digit at the last required decimal place. "
+      + "If the digit to the right of the last required decimal place is equidistant (i.e. a 5) "
+      + "then it is rounded up.";
+
+  private static final MathContext SIMPLE_ROUND_HALF_UP =
+      new MathContext(1, RoundingMode.HALF_UP);
+
+  @Udf
+  public Long round(@UdfParameter final Double val) {
+    return val == null ? null : Math.round(val);
+  }
+
+  @Udf
+  public Double round(@UdfParameter final Double val, @UdfParameter final Integer decimalPlaces) {
+    // Note that we MUST BigDecimal.valueOf(double) to create the BD to ensure correctness
+    // as this methods initialises via the string representation
+    return val == null
+        ?
+        null : roundBigDecimal(BigDecimal.valueOf(val), decimalPlaces).doubleValue();
+  }
+
+  @Udf(schemaProvider = "provideDecimalSchema")
+  public BigDecimal round(@UdfParameter final BigDecimal val) {
+    return val == null ? null : round(val, 0);
+  }
+
+  @Udf(schemaProvider = "provideDecimalSchemaWithDecimalPlaces")
+  public BigDecimal round(
+      @UdfParameter final BigDecimal val,
+      @UdfParameter
+      final Integer decimalPlaces
+  ) {
+    return val == null ? null : roundBigDecimal(val, decimalPlaces);
+  }
+
+  @UdfSchemaProvider
+  public SqlType provideDecimalSchemaWithDecimalPlaces(final List<SqlType> params) {
+    if (params.size() != 2) {
+      throw new KsqlException("Round udf with decimal places accepts two parameters");
+    }
+    final SqlType s0 = params.get(0);
+    if (s0.baseType() != SqlBaseType.DECIMAL) {
+      throw new KsqlException("The schema provider method for round expects a BigDecimal parameter"
+          + "type as first parameter.");
+    }
+    final SqlType s1 = params.get(1);
+    if (s1.baseType() != SqlBaseType.INTEGER) {
+      throw new KsqlException("The schema provider method for round expects an Integer parameter"
+          + "type as second parameter.");
+    }
+    return s0;
+  }
+
+  @UdfSchemaProvider
+  public SqlType provideDecimalSchema(final List<SqlType> params) {
+    if (params.size() != 1) {
+      throw new KsqlException("Round udf accepts one parameter");
+    }
+    final SqlType s0 = params.get(0);
+    if (s0.baseType() != SqlBaseType.DECIMAL) {
+      throw new KsqlException("The schema provider method for round expects a BigDecimal parameter"
+          + "type as a parameter.");
+    }
+    return s0;
```

Comment:
Q: should we be returning a decimal with zero scale here?  e.g. if input was `DECIMAL(5, 3)`, should we be returning a type of `DECIMAL(5,0)`, or `DECIMAL(2,0)` or `DECIMAL(5,3)`?

Given that `ROUND` is explicitly rounding the decimal, I would argue that `ROUND(123.123)`  is not `123.00` which is what the result will be if the scale is maintained. I would argue that the correct decimal answer is `123`, which has a scale of 0 and a precision of the input minus the scale.

I can see for the case of `ROUND` with two parameters we can't adjust the decimal type as we don't know what scale will be passed, and indeed the passed scale may vary from call to call. Hence the only thing we can do is stay with the same scale + precision.  That is not the case here. We know the scale is zero.  So maybe we should adjust the return type.

Thoughts?

Reply:
done